### PR TITLE
[clang-format] Adjust requires clause wrapping (#101550)

### PR DESCRIPTION
**Pull request by urnathan**
Address #101550 by adjusting line wrapping after a requires clause to permit a following '{' on the same line with OwnLine and WithPreceeding RequiresClausePositioning.

Thus, instead of:
```
bool Foo ()
  requires(true)
{
  return true;
}
```

we have:
```
bool Foo ()
  requires(true) {
  return true;
}
```

If the function body is empty, we'll get:
```
bool Foo ()
  requires(true) {}
```

I attempted to get a line break between the open and close braces, but failed. Perhaps that's fine -- it's rare and only happens in the empty body case.

### CODE REVIEW

**Comment by mydeveloperday** [comment]
This is changing tests, which means someone somewhere will call this a regression. Does this need an option in case one likes the existing style?

**Review by HazardyKnusperkeks — CHANGES_REQUESTED** [review]
Github says I have to comment something...

**Review by mydeveloperday — CHANGES_REQUESTED** [review]
I'm of the opinion if you are changing tests then something is wrong.

**Comment by urnathan** [comment]
> But that would be the way to go, add an option to the enum, don't change the existing behavior.

Sure -- I thought I'd shoot for the simplest change first :)

Here's an update that adds `OwnLineWithBrace`, which is like `OwnLine` except it also allows a trailing `{` (and anything that would then enable).  I describe the formatting in terms of the syntax (`{`) rather than semantics (function definition), to future proof it, should perhaps requires clauses be permitted at other brace-preceding locations.

File: clang/lib/Format/TokenAnnotator.cpp [hunk 1]
```diff
@@ -5682,12 +5683,15 @@ bool TokenAnnotator::mustBreakBefore(const AnnotatedLine &Line,
...
-      Right.isNot(tok::l_brace)) {
+  if (Left.ClosesRequiresClause && Right.isNot(tok::semi)) {
     switch (Style.RequiresClausePosition) {
     case FormatStyle::RCPS_OwnLine:
     case FormatStyle::RCPS_WithPreceding:
       return true;
+    case FormatStyle::RCPS_OwnLineWithBrace:
+      if (Right.isNot(tok::l_brace))
+        return true;
+      break;
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/TokenAnnotator.cpp:5694]:
```suggestion
      return Right.isNot(tok::l_brace);
```

When it's not a brace you want it to break.

I'm sure you can find a test case which fails without this change and passes with.

File: clang/unittests/Format/FormatTest.cpp [hunk 2]
```diff
@@ -25831,19 +25843,108 @@ TEST_F(FormatTest, RequiresClausesPositions) {
...
+               "{\n"
                "  return 5;\n"
                "}",
                Style);
 
+  Style.RequiresClausePosition = FormatStyle::RCPS_OwnLineWithBrace;
+  Style.IndentRequiresClause = true;
+
+  // The default in LLVM style is REI_OuterScope, but these tests were written
+  // when the default was REI_Keyword.
```
Comment (HazardyKnusperkeks) [inline clang/unittests/Format/FormatTest.cpp:25864]: Why?

Sure you can test the combination, but you always have to test LLVM + that single option.

File: clang/unittests/Format/FormatTest.cpp [hunk 3]
```diff
@@ -25831,19 +25843,108 @@ TEST_F(FormatTest, RequiresClausesPositions) {
...
                "int bar(T t)\n"
-               "requires F<T> {\n"
+               "requires F<T>\n"
+               "{\n"
                "  return 5;\n"
                "}",
                Style);
 
+  Style.RequiresClausePosition = FormatStyle::RCPS_OwnLineWithBrace;
+  Style.IndentRequiresClause = true;
```
Comment (HazardyKnusperkeks) [inline clang/unittests/Format/FormatTest.cpp:25861]: While we're at it. Please update the documentation of `IndentRequiresClause`.

File: clang/lib/Format/TokenAnnotator.cpp (hunk 1 above)
Comment (urnathan) [reply clang/lib/Format/TokenAnnotator.cpp]: I originally made the return conditional to match the ';' control flow.  But sure, direct return in both cases works too.

**Comment by urnathan** [comment]
Thanks for the comments, I think this addresses them

**Review by HazardyKnusperkeks — APPROVED** [review]

**Comment by urnathan** [comment]
@mydeveloperday, have I addressed your request -- I think so, as tests are no longer being changed.

**Review by mydeveloperday — APPROVED** [review]

**Pull request merged**